mountpoint-s3/tests/reftests/harness.rs

```diff
@@ -489,28 +557,32 @@ impl Harness {
     }
 
     async fn compare_file<'a>(&'a self, fs_file: InodeNo, ref_file: &'a MockObject) {
+        struct ReadVerifier(Box<[u8]>);
+
+        impl ReadReplier for ReadVerifier {
+            type Replied = ();
+
+            fn data(self, data: &[u8]) -> Self::Replied {
+                assert_eq!(&self.0[..], data, "read bytes don't match");
+            }
+
+            fn error(self, error: libc::c_int) -> Self::Replied {
+                panic!("read failed: {error}");
+            }
+        }
+
         let fh = self.fs.open(fs_file, 0x8000).await.unwrap().fh;
         let mut offset = 0;
         const MAX_READ_SIZE: usize = 4_096;
         let file_size = ref_file.len();
         while offset < file_size {
-            let mut read = Err(0);
             let num_bytes = MAX_READ_SIZE.min(file_size - offset);
+            let ref_bytes = ref_file.read(offset as u64, num_bytes);
+            assert_eq!(ref_bytes.len(), num_bytes);
+            let read_verifier = ReadVerifier(ref_bytes);
             self.fs
-                .read(
-                    fs_file,
-                    fh,
-                    offset as i64,
-                    num_bytes as u32,
-                    0,
-                    None,
-                    ReadReply(&mut read),
-                )
+                .read(fs_file, fh, offset as i64, num_bytes as u32, 0, None, read_verifier)
```

Comment:
This is an unrelated change but the performance was starting to annoy me while testing this PR. The idea here is just that `ReadReplier::data` doesn't need to copy every byte we read, which the old `ReadReply` struct was doing because it wanted to take ownership of the bytes.